### PR TITLE
feature: GCE image support

### PR DESCRIPTION
**Pull request by bentsi**
Changes:

Refactor `ScyllaAmiConfigurator` class to become general for AWS and GCE
- renamed to ScyllaMachineImageConfigurator
- support getting `endpoint_snitch` from instance based in `scylla_util.py`
- support getting instance metadata using [aws|gce]_instance class based in `scylla_util.py`
- support getting user data from instance metadata in GCE

Refactor `scylla_image_setup`
- generalize the script to run on both GCE image and AMI
- move to `common` dir
- use `io_setup` method from an instance from `scylla_util.py` based classes

Refactor scylla_login 
- generalize the script to run on both GCE image and AMI
- move to `common` dir
- list of optimized instances list is taken from `GETTING_STARTED_URL` attribute of the relevant instance class in  `scylla_util.py`
- added Scylla version to the login screen
- replaces AMI by the "image" word
- running a check method on instance class from the relevant instance object defined in `scylla_util.py`

Improvements to `dist/redhat/build_rpm.sh`
- running tests from the new path
- added centos8 target
- moved the AWS related code to the if statement 

RPM spec file
- changed paths to the scripts that were moved to other locations

- Added Build image script for GCE
- Created `scylla_install_image`  script specific for GCE
- new `scylla.json` used by Packer to create GCE image
- GCE adapted version of scylla_create_devices
- Refactor and generalize `test_scylla_configure.py` based on the changes to `ScyllaAmiConfigurator`

Travis:
-  added a test to build GCE rpm

### CODE REVIEW

**Comment by roydahan** [comment]
@fruch can you please review this PR.

**Comment by slivne** [comment]
@penberg please make sure that either you or taluya review this pr this week - so we can move forward and have a gce image

@fruch please review as well

File: aws/scylla_create_devices
Comment (fruch) [inline aws/scylla_create_devices:29]: why are we removing logging for this one ?

**Comment by fruch** [comment]
@bentsi  you might want to comment here: this is depended on: https://github.com/scylladb/scylla/pull/7080/

**Comment by fruch** [comment]
@bentsi Did we created AMI out of this branch ?

**Review by fruch — APPROVED** [review]
Looks good,
* should make sure we merge https://github.com/scylladb/scylla/pull/7080/ first, 
* make sure we test AMI creating, to see it wasn't broken.

**Comment by fruch** [comment]
@penberg who's going to do the scylla-pkg part ? creating a job and promoting ?

**Comment by penberg** [comment]
@slivne I am not sure what you want me to review: there's no description of what is being changed and why. What I am able to reverse engineer from the diff looks like a reasonable approach, but there's no way for me to give any meaningful comments about it.

File: aws/scylla_create_devices
Comment (penberg) [reply aws/scylla_create_devices:29]: @bentsi I am also curious -- why is logging being removed? What's the impact of this to existing systems?

**Comment by roydahan** [comment]
@bentsi can you please elaborate in the commit message?

**Comment by slivne** [comment]
@bentsi ping

**Comment by bentsi** [comment]
update description, @penberg hopefully it will help you understand the changes better

**Comment by penberg** [comment]
@bentsi Where can I find "scylla_util" and specifically the `get_cloud_instance` function?

**Comment by penberg** [comment]
@bentsi Never mind, it's in `scylla.git`...

File: common/scylla_configure.py
Comment (penberg) [inline common/scylla_configure.py:32]: If the configurator is now generic, shouldn't `AMI_CONF_DEFAULTS` be renamed to something else?

File: common/scylla_configure.py
Comment (penberg) [inline common/scylla_configure.py:51]: We should rename the file to something else, no? But let's also keep `ami_disabled` for backward compatibility.

File: aws/scylla_create_devices
Comment (penberg) [reply aws/scylla_create_devices:29]: @bentsi, ping?

File: common/scylla_configure.py
Comment (penberg) [inline common/scylla_configure.py:65]: No need to address in this pull request, but I would like to understand why cloud provider support is in `scylla.git`, not in this repository...

**Review by penberg — CHANGES_REQUESTED** [review]

File: common/scylla_configure.py
Comment (fruch) [reply common/scylla_configure.py:65]: I guess since `scylla_setup` still need to use this logic (and historically it was there), anyhow both repos need those classes to understand in which cloud they are, and get some basic info out of it.

File: common/scylla_configure.py
Comment (fruch) [reply common/scylla_configure.py:51]: This name can't be change without changing it in code, this is the file that stop scylla service for starting.

It can be changed, but it need to be in sync with scylla-core.

**Comment by slivne** [comment]
@bentsi ping

File: common/scylla_configure.py
Comment (bentsi) [reply common/scylla_configure.py:51]: yeah, I didn't want to change the legacy code, that's why I kept this. We can do this in the future as an improvement.

File: aws/scylla_create_devices
Comment (bentsi) [reply aws/scylla_create_devices:29]: `scylla_create_devices` is a simple script that can be run outside the scylla-machine-image setup, it shouldn't use the Scylla Machine Image (SMI) log. The log was created for SMI configurator and not for the scylla-image-setup.service (it uses the systemd facility)

File: common/scylla_configure.py
Comment (bentsi) [reply common/scylla_configure.py]: fixed.

**Comment by bentsi** [comment]
addressed all the comments.

**Review by fruch — APPROVED** [review]
LGTM

**Comment by fruch** [comment]
@bentsi one question I didn't saw the answer, did we built AWS AMI using this code ? to see we didn't broke that ?

**Comment by bentsi** [comment]
@fruch that's a good question, that I wanted to address. This is change is potentially a breaking change for the AMI Jenkins pipeline.
I am not familiar with it but some paths changed here. @penberg can you ask somebody in your team or point me to the job that creates the AMI so I can use it to test this PR?

File: gce/scylla_create_devices
Comment (zimnx) [inline gce/scylla_create_devices:33]: Can you parametrize --raiddev and --root? Simillar to https://github.com/scylladb/scylla-machine-image/pull/50/files#diff-6e97616a1e8944b01fcc6ef68257e0d6
This script will be used by the other PR once both of them will get merged.

**Comment by slivne** [comment]
@zimnx we can do this in followups we don't need this for merge - unless its simple

@penberg there is an open request for help from you  to help fix an item.

File: common/scylla_configure.py
Comment (penberg) [reply common/scylla_configure.py:51]: Please open a ticket for this cleanup so we don't forget.

**Comment by yaronkaikov** [comment]
> @fruch that's a good question, that I wanted to address. This is change is potentially a breaking change for the AMI Jenkins pipeline.
> I am not familiar with it but some paths changed here. @penberg can you ask somebody in your team or point me to the job that creates the AMI so I can use it to test this PR?

@bentsi https://jenkins.scylladb.com/job/scylla-master/job/ami/ That's the job creating the AMI for master. if you want to debug you can either copy it, or use https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/yaron-debug/job/ami/

**Comment by slivne** [comment]
@bentsi ping

**Comment by bentsi** [comment]
@yaronkaikov I can't find the place in the job where I can set custom branch and repo for scylla-machine-image...

**Comment by yaronkaikov** [comment]
@bentsi Sorry i misunderstood what you need, this is the job you should use https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/yaron-debug/job/byo_build_tests_dtest/ - if you need my assistance, please contact me tomorrow morning

**Comment by bentsi** [comment]
@yaronkaikov let's discuss tomorrow morning.

**Comment by yaronkaikov** [comment]
@bentsi So during the first test we discussed i found a bug in scylla-pkg. fixed it, waiting for review to get into next https://github.com/scylladb/scylla-pkg/pull/1614

**Review by yaronkaikov — APPROVED** [review]
It's not braking the current AMI build

**Comment by fruch** [comment]
> It's not braking the current AMI build

@yaronkaikov Can you link the job that succeeded for completeness ?

**Comment by yaronkaikov** [comment]
> > @fruch that's a good question, that I wanted to address. This is change is potentially a breaking change for the AMI Jenkins pipeline.
> 
> > I am not familiar with it but some paths changed here. @penberg can you ask somebody in your team or point me to the job that creates the AMI so I can use it to test this PR?
> 
> 
> 
> @bentsi https://jenkins.scylladb.com/job/scylla-master/job/ami/ That's the job creating the AMI for master. if you want to debug you can either copy it, or use https://jenkins.scylladb.com/job/scylla-master/job/releng-testing/job/yaron-debug/job/ami/

👆

**Comment by yaronkaikov** [comment]
> @bentsi So during the first test we discussed i found a bug in scylla-pkg. fixed it, waiting for review to get into next [scylladb/scylla-pkg#1614](https://github.com/scylladb/scylla-pkg/pull/1614)
> 
> I have manage to build AMI using my branch (for the fix) and your branch and it was created successfully - https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/yaron-debug/job/byo-package/20/
> 
> So you changes doesn't seems to break anything

@fruch ^^

**Comment by bentsi** [comment]
@yaronkaikov 
Looks like the job used olds version of SMI:
00:05:03  scylla-machine-image-666.development-20200805.f2e30f54a6e.noarch.rpm

**Comment by yaronkaikov** [comment]
@bentsi Are you looking at the right logs? https://jenkins.scylladb.com/view/master/job/scylla-master/job/releng-testing/job/yaron-debug/job/byo-package/20/consoleFull - looks ok to me

**Comment by bentsi** [comment]
I looked in the wrong job. I see that it used the new SMI version, I will run AMI test on the created Image and update here.

**Comment by fruch** [comment]
> I looked in the wrong job. I see that it used the new SMI version, I will run AMI test on the created Image and update here.
@yaronkaikov @bentsi 
can we spin a AMI test, just for one last check ?

**Comment by bentsi** [comment]
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/bentsi-staging/job/bentsi-artifacts-ami/1/console
AMI test passed. So this PR is not breaking current AMI.

**Comment by yaronkaikov** [comment]
@bentsi So you want to merge, then i will start my work on pipeline

**Review by fruch — APPROVED** [review]
Now that I have my proof, AWS AMI aren't broken.
LGTM

**Pull request merged**
